docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md

```diff
@@ -0,0 +1,412 @@
+assembly that defines the ActivitySource, not the assembly whose code is being instrumented.
+
+- The version parameter is optional. It is recommended to provide the version in case you release multiple
+versions of the library and make changes to the instrumented telemetry.
+
+> [!NOTE]
+> OpenTelemetry uses alternate terms 'Tracer' and 'Span'. In .NET 'ActivitySource' is the implementation
+> of Tracer and Activity is the implementation of 'Span'. .NET's Activity type long pre-dates
+> the OpenTelemetry specification and the original .NET naming has been preserved for
+> consistency within the .NET ecosystem and .NET application compatibility.
+
+### Activity Creation
+
+Use the ActivitySource object to Start and Stop Activity objects around meaningful units of work. Update
+DoSomeWork() with the code shown here:
+
+```csharp
+        static async Task DoSomeWork(string foo, int bar)
+        {
+            using (Activity activity = source.StartActivity("SomeWork"))
+            {
+                await StepOne();
+                await StepTwo();
+            }
+        }
+```
+
+Running the app now shows the new Activity being logged:
+```dotnetcli
+> dotnet run
+dotnet run
+Activity.Id:          00-9a41c8f2144cef45a99b9164db9da0b6-5cbc0f6836635245-01
+Activity.DisplayName: SomeWork
+Activity.Kind:        Internal
+Activity.StartTime:   2021-03-15T08:39:24.7379085Z
+Activity.Duration:    00:00:01.5198773
+Resource associated with Activity:
+    service.name: unknown_service:temp
+
+Example work done
+```
+
+#### Notes
+
+- <xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType> creates and starts
+the activity at the same time. The listed code pattern is using the `using` block which automatically disposes
+the created Activity object after executing the block. Disposing the Activity object will stop it so the code
+doesn't need to explicitly call <xref:System.Diagnostics.Activity.Stop?displayProperty=nameWithType>.
+That simplifies the coding pattern.
+
+- <xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType> internally determines if
+there are any listeners recording the Activity. If there are no registered listeners or there are listeners which
+are not interested, `ActivitySource.StartActivity` will return `null` and avoid creating the Activity object. This
+is a performance optimization so that the code pattern can still be used in functions that are called very
+frequently.
+
+## Optional: Populate tags
+
+Activities support key-value data called Tags, commonly used to store any parameters of the work that
+may be useful for diagnostics. Update DoSomeWork() to include them:
+
+```csharp
+        static async Task DoSomeWork(string foo, int bar)
+        {
+            using (Activity activity = source.StartActivity("SomeWork"))
+            {
+                activity?.SetTag("foo", foo);
```

Comment:
Do we want to encourage people to set simple tags in the `StartActivity` call (to avoid extra API invocations) and only use `activity?.SetTag` for tags that are expensive to evaluate?

Reply:
I'm less worried about API invocations as I am allocations. Specifying tags in StartActivity requires allocating an IEnumerable and the allocation still happens even for unsampled Activities. I'm glad to have us change the guidance if we decide specifying them in StartActivity() is preferable, but so far all examples I recall (including the [OpenTelemetry ones](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started/Program.cs#L36)) use the SetTag() pattern.

Reply:
Otel spec recommendation:
"Whenever possible, users SHOULD set any already known attributes at span creation instead of calling SetAttribute later."
https://github.com/open-telemetry/opentelemetry-specification/blob/v1.0.1/specification/trace/api.md#span-creation

But agree, its *not* free in .NET as it requires alloc before knowing sampling decision.

Reply:
Do you have any background on that spec recommendation? I could guess it is designed to make it more reliable for samplers to sample using attributes but I've never seen an example of any sampler that cared about attributes. It makes me sad to trade away performance in cases that apps aren't even logging telemetry for a scenario that appears to be rare?

Reply:
I have seen people asking about this and some writing custom sampler (which wraps the default ones) to sample out certain paths based on known attributes.

(Btw, I try to interpret it as - if the attributes are readily available (i.e available already as IEnumerable for whatever other reasons - like metric/log purpose, then use it in Activity creation, since this should be relatively free.)

Reply:
yeah if the attributes are already in an IEnumerable that would be great, though I am suspicious that is going to be the rare case.